lib/Logging.js

```diff
@@ -0,0 +1,149 @@
+const winston = require("winston");
+require('winston-daily-rotate-file');
+
+class Logging {
+    constructor() {
+        this.loggers = new Map();
+        this.transports = null;
+    }
+
+    /*
+        config:
+            console: "error|warn|info|debug|off"
+            files: {
+                "abc.log" => "error|warn|info|debug|off"
+            }
+            maxFiles: 5
+    */
+    Configure(config) {
+        let updatingLoggers = false;
+        const timestampFn = function() {
+            return new Date().toISOString().replace(/T/, ' ').replace(/\..+/, '');
+        };
+        const formatterFn = function(opts) {
+            return opts.timestamp() + ' ' +
+            opts.level.toUpperCase() + ':' +
+            (opts.meta && opts.meta.loggerName ? opts.meta.loggerName : "") + ' ' +
+            (undefined !== opts.message ? opts.message : '');
+        };
+        // if (this.transports.length > 0) {
+        //     updatingLoggers = true;
+        //     this.transports.forEach((transport) => {
+        //         this.loggers.forEach((wrapper, name) => {
+        //             wrapper.logger.remove(transport);
+        //         });
+        //     });
+        // }
+        this.transports = [];
+        if (config.console != undefined && config.console != "off") {
+            this.transports.push(new (winston.transports.Console)({
+                json: false,
+                name: "console",
+                timestamp: timestampFn,
+                formatter: formatterFn,
+                level: config.console
+            }));
+        }
+
+        if(config.files == undefined) {
+            return;
+        }
+
+        let i = 0;
+        for (let file of config.files) {
+            const filename = Object.keys(file)[0];
+            const level = file[filename];
+            i++;
+            this.transports.push(new (winston.transports.DailyRotateFile)({
+                filename,
+                datePattern: "YYYY-MM-DD",
+                name: `logfile` + i,
+                formatter: formatterFn,
+                level,
+                timestamp: timestampFn,
+                maxFiles: config.maxFiles > 0 ? config.maxFiles : undefined
+            }));
+        }
+
+        this.loggers.forEach((wrapper, name) => {
+            wrapper.SwapIn(this.createLogger(name));
+        });
+    }
+
+    Get(name) {
+        if(!this.loggers.has(name)) {
+            const wrapper = new LogWrapper()
+            this.loggers.set(name, wrapper);
+            if (this.transports != null) {
```

Comment:
As above

Reply:
This is also quite unclear, why do we set the `logger` of `wrapper` if `this.transports != null` ?

Reply:
The wrapper is there to ensure messages aren't lost if we've not setup the transports yet. We create the wrapper regardless, but if the transports *are* ready we also create the actual logger.